lib/spack/spack/modules/common.py

```diff
@@ -269,13 +269,39 @@ def read_module_indices():
 module_type_to_indices = read_module_indices()
 
 
-def upstream_module(spec, module_type):
-    indices = module_type_to_indices[module_type]
+def upstream_module(spec, module_type, _module_indices=None):
+    indices = _module_indices or module_type_to_indices[module_type]
     for index in indices:
         if spec.dag_hash() in index:
             return index[spec.dag_hash()]
 
 
+def get_module(module_type, spec, get_full_path, module_indices=None):
+    if spec.package.installed_upstream:
```

Comment:
If the package is also installed locally (`spec.package.installed`), shouldn't that take precedence?

Reply:
Yes: `installed_upstream` is `False` if the package is installed locally (even if it is also installed upstream)

Reply:
I think this may end up being confusing one day, but I am not sure how best to handle it.  I thought about this and wondered to myself whether we should even leverage upstream modules -- or whether we should just generate replacements for them, and I could see users doing either for good reason.

I think we may need some way to talk about upstream modules even if something is installed locally.  You can imagine a case where something is installed BOTH upstream and locally but a package still remains that was built with the upstream version.  But I think this can be refined in a later PR.

Reply:
The only case where a package would be installed both locally and upstream, is where the local package was installed first. In that case, I think the local package should always be used, because the RPATHs in the package point to the local copy (and for that reason I think the module you get in that case should also be for the local install); is there a case that I'm missing (or does that seem otherwise incomplete)?